asynction/security.py

```diff
@@ -0,0 +1,391 @@
+            )
+
+        return token_info
+
+    return http_api_key_security_check
+
+
+def build_oauth2_security_check(
+    requirement: SecurityRequirement, scheme: SecurityScheme
+) -> Callable[[Request], Mapping]:
+    token_info_func = load_token_info_func(scheme)
+    required_scopes = requirement.scopes
+
+    def oauth2_security_check(request: Request):
+        token_info = validate_authorization_header(request, token_info_func)
+        scope_validate_func = load_scope_validate_func(scheme)
+
+        if token_info is None:
+            return None
+
+        token_scopes = token_info.get("scope", token_info.get("scopes", ""))
+
+        if not scope_validate_func(required_scopes, token_scopes):
+            raise SecurityException(
+                f"Invalid scopes: required: {required_scopes}, provided: {token_scopes}"  # noqa: 501
+            )
+
+        return token_info
+
+    return oauth2_security_check
+
+
+# Dispatch table mapping SecuritySchemesType to security check builder
+_BUILDER_DISPATCH = {
+    SecuritySchemesType.HTTP: build_http_security_check,
+    SecuritySchemesType.OAUTH2: build_oauth2_security_check,
+    SecuritySchemesType.HTTP_API_KEY: build_http_api_key_security_check,
+}
+
+
+def build_security_handler(
+    security: Sequence[SecurityRequirement],
+    security_schemes: Mapping[str, SecurityScheme],
+) -> Callable[[Request], Mapping]:
+    # build a list of security validators based on the provided security schemes
+    security_checks = []
+
+    for requirement in security:
+        scheme = security_schemes.get(requirement.name)
+        if not scheme:
+            raise UnregisteredSecurityScheme(requirement.name)
+
+        scheme_type = SecuritySchemesType(scheme.type)
+        builder = _BUILDER_DISPATCH.get(scheme_type)
+        if not builder:
+            raise UnsupportedSecurityScheme(str(scheme_type))
+
+        security_checks.append(builder(requirement, scheme))
+
+    def security_handler(request: Request) -> Mapping:
+
+        try:
+            # apply the security schemes in the order listed in the API file
+            for check in security_checks:
+
+                # if a security check fails if will raise the appropriate exception
+                # if the security check passes it will return a dict of kwargs to pass to the handler   # noqa: 501
+                # if the check is not applicable based on lack provided argument the check will return None indicating   # noqa: 501
+                # that the next (if any) check should be run.
+                security_args = check(request)
+                if security_args:
+                    return security_args
+            else:
+                raise SecurityException("No checks passed")
+        except SecurityException as err:
+            print(err, file=sys.stdout)
+            raise ConnectionRefusedError(str(err)) from err
```

Comment:
```suggestion
            raise SecurityException("Failed security checks")  # would be good to have scheme type information within the exception message
```
We should raise an asynction error here (probably SecurityException) so that users can distinguish it between any other non-asynction error. Maybe the `check` method could return the args and a potentioal err message:
```python
args, err = check(request)
if err:
    raise SecurityException(f"Security check failed: {err}")
return args
```